wip: add integration testing guidelines (#6)
tier4/autoware-documentation: docs/contributing/testing-guidelines/integration-testing.md

# Integration Testing

> ## Introduction

Reviewer: I think you can write an overview right below the title.

```suggestion
```

> This article motivates developers to adopt integration testing by explaining how to write, run, and evaluate the results of integration tests.

Reviewer: Let's read and follow the Google developer documentation style guide as much as possible. :+1: 
https://developers.google.com/style/translation?hl=en#write-short,-clear,-and-precise-sentences

In this case, I think we can write like this:
```suggestion
This article explains how to write integration tests.
```

Reviewer: See also: https://github.com/autowarefoundation/autoware/pull/14

## Integration testing

An integration test is defined as the phase in software testing where individual software modules are combined and tested as a group. Integration tests occur after unit tests, and before validation tests.

The input to an integration test is a set of independent modules that have been unit tested. The set of modules are tested against the defined integration test plan, and the output is a set of properly integrated software modules that are ready for system testing.

## Value of integration testing

Integration tests determine if independently developed software modules work correctly when the modules are connected to each other. In ROS 2, the software modules are called nodes. As a special case, testing a single node can be referred to as component testing.

Integration tests help to find the following types of errors:

- Incompatible interaction between nodes, such as non-matching topics, different message types, or incompatible QoS settings
- Reveal edge cases that were not touched with unit tests, such as a critical timing issue, network communication delay, disk I/O failure, and many other problems that can occur in production environments

> - Using tools like `stress` and `udpreplay`, performance of nodes is tested with real data or while the system is under high CPU/memory load, where situations such as `malloc` failures can be detected

Reviewer: https://github.com/tier4/autoware-spell-check-dict/pull/245

In Autoware.Core, we use the [launch_testing](https://github.com/ros2/launch/tree/master/launch_testing) framework.

### Smoke tests

Autoware has dedicated API for smoke testing

To use this framework, in `package.xml` add:

```xml
<test_depend>autoware_testing</test_depend>
```

and in `CMakeLists.txt` add:

```cmake
if(BUILD_TESTING)
  find_package(autoware_testing REQUIRED)
  add_smoke_test(${PROJECT_NAME} ${NODE_NAME})
endif()
```

which adds smoke test that ensures that node can be:

1. launched with default parameter file,
2. terminated with a standard `SIGTERM` signal,

For full API documentation see [package design page](https://github.com/autowarefoundation/autoware.universe/blob/main/common/autoware_testing/design/autoware_testing-design.md).

This API is not suitable for all smoke test cases. For example, it can not be used when some specific file location, like map, is required to be passed to the node or some preparation need to be conducted before node launch. In such cases use manual solution from [section below](#integration-test-with-a-single-node-component-test).

> ### Interface tests

Author: この項目は新規追加しました。

Reviewer: それは良さそうですが、テストの全体像を先に書く必要がありそうです。
別のPRにありましたっけ？

Reviewer: （これか https://github.com/autowarefoundation/autoware.universe/pull/391）
こちらもautoware-documentationに持ってきた方が良さそうです。

        executable='lanelet2_map_provider_exe',
        namespace='had_maps',
        parameters=[
            os.path.join(
                get_package_share_directory('lanelet2_map_provider'),
                'param/test.param.yaml'
            ),
            {
                'map_osm_file': map_osm_file
            }]
    )

    context = {'lanelet2_map_provider': lanelet2_map_provider}

    return LaunchDescription([
        lanelet2_map_provider,
        # Start tests right away - no need to wait for anything
        launch_testing.actions.ReadyToTest()]
    ), context
```

and finally the test condition. As before, it is just a smoke test ensures the node can be

1. launched with its default parameter file,
2. terminated with a standard `SIGTERM` signal,

so the test code is executed after the node executable has been shut down (`post_shutdown_test`):

```python
@launch_testing.post_shutdown_test()
class TestProcessOutput(unittest.TestCase):

    def test_exit_code(self, proc_output, proc_info, ndt_mapper):
        # Check that process exits with code 0
        launch_testing.asserts.assertExitCodes(proc_info, process=ndt_mapper)
```

> ## Running the test

Reviewer: Let's move the descriptions about how to run tests to another page.
We need to keep pages short and simple.

#### Testing the output of a node

To test while the node is running, create an [_active test_](https://github.com/ros2/launch/tree/foxy/launch_testing#active-tests) by adding a subclass of Python's `unittest.TestCase` to `*launch.test.py`. Some boilerplate code is required to access output by creating a node and a subscription to a particular topic; e.g.

```python
import unittest

class TestRunningDataPublisher(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls.context = Context()
        rclpy.init(context=cls.context)
        cls.node = rclpy.create_node("test_node", context=cls.context)

    @classmethod
    def tearDownClass(cls):
        rclpy.shutdown(context=cls.context)

    def setUp(self):
        self.msgs = []
        sub = self.node.create_subscription(
            msg_type=my_msg_type,
            topic="/info_test",
            callback=self._msg_received
        )
        self.addCleanup(self.node.destroy_subscription, sub)

    def _msg_received(self, msg):
        # Callback for ROS 2 subscriber used in the test
        self.msgs.append(msg)

    def get_message(self):

>         startlen = len(self.msgs)

Reviewer: ```suggestion
        start_len = len(self.msgs)
```
